File: 21/b.ts
// const input = await Bun.file("test.txt").text();
const input = await Bun.file("input.txt").text();

let startY = 0;
let startX = 0;

const board = input.split("\n").map((d, y) => {
  return d.split("").map((d, x) => {
    if (d === "#") {
      return true;
    }
    if (d === "S") {
      startY = y;
      startX = x;
    }
    return false;
  });
});

let res = [];

let canReach: [number, number][] = [[startX, startY]];

const test = (x: number, y: number) => {
  x %= board[0].length;
  y %= board.length;
  if (x < 0) x += board[0].length;
  if (y < 0) y += board.length;
  return !board[y][x];
};

let step = 1;

while (true) {
  console.log(step, canReach.length);
  const newCanReach: [number, number][] = [];
  for (const [x, y] of canReach) {
    if (
      test(x - 1, y) &&
      !newCanReach.some((d) => d[0] === x - 1 && d[1] === y)
    ) {
      newCanReach.push([x - 1, y]);
    }
    if (
      test(x + 1, y) &&
      !newCanReach.some((d) => d[0] === x + 1 && d[1] === y)
    ) {
      newCanReach.push([x + 1, y]);
    }
    if (
      test(x, y - 1) &&
      !newCanReach.some((d) => d[0] === x && d[1] === y - 1)
    ) {
      newCanReach.push([x, y - 1]);
    }
    if (
      test(x, y + 1) &&
      !newCanReach.some((d) => d[0] === x && d[1] === y + 1)
    ) {
      newCanReach.push([x, y + 1]);
    }
  }
  if (step === Math.floor(0.5 * board.length) + res.length * board.length) {
    res.push(newCanReach.length);
    if (res.length === 3) {
      let n = 26501365 / board.length;

      let a0 = res[0];
      let a1 = res[1];
      let a2 = res[2];

      let b0 = a0;
      let b1 = a1 - a0;
      let b2 = a2 - a1;
      console.log(b0 + b1 * n + ((n * (n - 1)) / 2) * (b2 - b1));
      break;
    }
  }
  step++;
  canReach = newCanReach;
}

File: 21/b.rs
use std::{collections::HashSet, fs};

use pathfinding::grid;

#[derive(Debug, Hash, Eq, PartialEq)]
struct Coord(i64, i64);

impl Coord {
  fn neighbours(&self) -> Vec<Self> {
    vec![
      Coord(self.0 + 1, self.1),
      Coord(self.0 - 1, self.1),
      Coord(self.0, self.1 + 1),
      Coord(self.0, self.1 - 1),
    ]
  }
}

fn main() {
  let file = fs::read_to_string("input.txt").unwrap();
  let mut start = Coord(0, 0);
  let board = file
    .lines()
    .enumerate()
    .map(|(y, line)| {
      line
        .chars()
        .enumerate()
        .map(|(x, char)| {
          if char == 'S' {
            start = Coord(x as i64, y as i64);
          }
          return char == '#';
        })
        .collect::<Vec<bool>>()
    })
    .collect::<Vec<Vec<bool>>>();

  let mut can_reach = HashSet::new();
  can_reach.insert(start);

  let mut res = Vec::new();

  for step in 1.. {
    println!("{} {}", step, can_reach.len());
    let mut new_can_reach = HashSet::new();
    for coord in can_reach.iter() {
      for n in coord.neighbours().into_iter().filter(|pos| {
        let y = pos.1.rem_euclid(board.len() as i64) as usize;
        let x = pos.0.rem_euclid(board.len() as i64) as usize;
        !board[y][x]
      }) {
        new_can_reach.insert(n);
      }
    }
    can_reach = new_can_reach;
    if step == board.len() / 2 + board.len() * res.len() {
      res.push(can_reach.len());
      if res.len() == 3 {
        break;
      };
    }
  }
  let n = 26501365 / board.len();

  let b0 = res[0];
  let b1 = res[1] - res[0];
  let b2 = res[2] - res[1];

  println!("{}", b0 + b1 * n + (n * (n - 1) / 2) * (b2 - b1))
}
